File: dsa-master/sort/mergeSort.js
const merge = (arr, l, mid, h) => {
  let left = l;
  let right = mid + 1;
  const newArr = [];
  while (left <= mid && right <= h) {
    if (arr[left] < arr[right]) {
      newArr.push(arr[left]);
      left++;
    } else {
      newArr.push(arr[right]);
      right++;
    }
  }

  while (left <= mid) {
    newArr.push(arr[left]);
    left++;
  }
  while (right <= h) {
    newArr.push(arr[right]);
    right++;
  }

  for (let i = 0; i <= newArr.length-1; i++) {
    arr[i+l] = newArr[i];
  }
};

const mergeSort = (arr, l, h) => {
  if (l == h) return;
  let mid = Math.floor((l + h) / 2);

  mergeSort(arr, l, mid);
  mergeSort(arr, mid + 1, h);

  merge(arr, l,mid, h);
};

const solve = () => {
  const arr = [8, 3, 7, 4, 9, 2, 6, 1,0];
  mergeSort(arr, 0, arr.length - 1);
  //   merge(arr, 0, arr.length - 1);

  for (let i = 0; i <= arr.length - 1; i++) {
    console.log(arr[i]);
  }
};

solve();

File: dsa-master/sort/mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int> &vct, int low, int mid, int high)
{
    int left = low;
    int right = mid + 1;
    vector<int> merged;
    while (low <= mid and right <= high)
    {
        if (vct[left] < vct[right])
        {
            merged.push_back(vct[left]);
            left++;
        }
        else
        {
            merged.push_back(vct[right]);
            right++;
        }
    }

    while (left <= mid)
    {
        merged.push_back(vct[left]);
        left++;
    }

    while (right <= high)
    {
        merged.push_back(vct[right]);
        right++;
    }

    for (int i = 0; i < merged.size(); i++)
    {
        vct[low + i] = merged[i];
    }
}

void mergeSort(vector<int> &vct, int low, int high)
{
    if (low == high)
        return;
    int mid = (low + high) / 2;

    mergeSort(vct, low, mid);
    mergeSort(vct, mid + 1, high);

    merge(vct, low, mid, high);
}

int main()
{

    vector<int> vct = {3, 2,5, 4, 1, 3};

    mergeSort(vct, 0, vct.size() - 1);

    for (auto it : vct)
    {
        cout << it << " ";
    }
    return 0;
}